osu.Game.Rulesets.Osu/Mods/OsuModBubbles.cs

```diff
@@ -0,0 +1,255 @@
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Objects.Drawables;
+using osu.Game.Rulesets.Osu.Objects;
+using osu.Game.Rulesets.Osu.Objects.Drawables;
+using osu.Game.Rulesets.Osu.Skinning.Default;
+using osu.Game.Rulesets.Scoring;
+using osu.Game.Rulesets.UI;
+using osu.Game.Scoring;
+using osuTK;
+
+namespace osu.Game.Rulesets.Osu.Mods
+{
+    public partial class OsuModBubbles : ModWithVisibilityAdjustment, IApplicableToDrawableRuleset<OsuHitObject>, IApplicableToScoreProcessor
+    {
+        public override string Name => "Bubbles";
+
+        public override string Acronym => "BU";
+
+        public override LocalisableString Description => "Don't let their popping distract you!";
+
+        public override double ScoreMultiplier => 1;
+
+        public override ModType Type => ModType.Fun;
+
+        // Compatibility with these seems potentially feasible in the future, blocked for now because they don't work as one would expect
+        public override Type[] IncompatibleMods => new[] { typeof(OsuModBarrelRoll), typeof(OsuModMagnetised), typeof(OsuModRepel) };
+
+        private PlayfieldAdjustmentContainer bubbleContainer = null!;
+
+        private readonly Bindable<int> currentCombo = new BindableInt();
+
+        private float maxSize;
+        private float bubbleRadius;
+        private double bubbleFade;
+
+        private readonly DrawablePool<BubbleDrawable> bubblePool = new DrawablePool<BubbleDrawable>(100);
+
+        private DrawableOsuHitObject lastJudgedHitobject = null!;
+
+        public ScoreRank AdjustRank(ScoreRank rank, double accuracy) => rank;
+
+        public void ApplyToScoreProcessor(ScoreProcessor scoreProcessor)
+        {
+            currentCombo.BindTo(scoreProcessor.Combo);
+            currentCombo.BindValueChanged(combo =>
+                maxSize = Math.Min(1.75f, (float)(1.25 + 0.005 * combo.NewValue)), true);
+
+            scoreProcessor.NewJudgement += result =>
+            {
+                if (result.HitObject is not OsuHitObject osuHitObject) return;
+
+                DrawableOsuHitObject drawableOsuHitObject = lastJudgedHitobject;
+
+                switch (result.HitObject)
+                {
+                    case Slider:
+                    case SpinnerTick:
+                        break;
+
+                    default:
+                        addBubble();
+                        break;
+                }
+
+                void addBubble()
+                {
+                    BubbleDrawable bubble = bubblePool.Get();
+                    bubble.Info = new BubbleInfo
+                    {
+                        InitialSize = new Vector2(bubbleRadius),
+                        MaxSize = maxSize,
+                        Position = getPosition(),
+                        FadeTime = bubbleFade,
+                        Colour = drawableOsuHitObject.AccentColour.Value,
+                        IsHit = drawableOsuHitObject.IsHit,
+                    };
```

Comment:
Isn't everything here that isn't already provided by `DrawableOsuHitObject` constant to all? I'm not sure this struct is required, and don't think it's adding any value.

Reply:
im not quite sure what you are proposing, that i feed the bubble its drawable hitobject directly and then make properties inside of it for anything that isn't sourced from the drawablehitobject?

Reply:
Basically yeah.